github/downloader_test.go

```diff
@@ -80,74 +109,378 @@ func getDownloader() (*Downloader, *testutils.Memory, error) {
-	if err != nil {
-		t.Errorf("Failed to read the testcases:%s", err)
-	}
-
+	tests, err := loadTests("../testdata/online-organization-tests.json")
+	suite.NoError(err, "Failed to read the testcases")
 	downloader, storer, err := getDownloader()
-	require.NoError(t, err)
-
-	for _, test := range tests.RepositoryTests {
-		t.Run(fmt.Sprintf("%s/%s", test.Owner, test.Repository), func(t *testing.T) {
-			testOnlineRepo(t, test, downloader, storer)
+	suite.NoError(err, "Failed to instantiate downloader")
+	for _, test := range tests.OrganizationsTests {
+		test := test
+		t.Run(fmt.Sprintf("Org: %s", test.Org), func(t *testing.T) {
+			testOrg(t, test, downloader, storer)
 		})
 	}
 }
 
-func testOnlineOrg(t *testing.T, oracle OrganizationTest, d *Downloader, storer *testutils.Memory) {
+func testOrg(t *testing.T, oracle testutils.OrganizationTest, d *Downloader, storer *testutils.Memory) {
 	err := d.DownloadOrganization(context.TODO(), oracle.Org, oracle.Version)
 	require := require.New(t)
 	require.Nil(err, "DownloadOrganization(%s) failed", oracle.Org)
-	// Sample some properties that will not change, no topics available in git-fixtures
-	require.Equal(oracle.Org, storer.Organization.Name)
+	require.Equal(oracle.Org, storer.Organization.Login)
 	require.Equal(oracle.URL, storer.Organization.Url)
 	require.Equal(oracle.CreatedAt, storer.Organization.CreatedAt.String())
 	require.Equal(oracle.PublicRepos, storer.Organization.PublicRepos.TotalCount)
 	require.Equal(oracle.TotalPrivateRepos, storer.Organization.TotalPrivateRepos.TotalCount)
 	require.Len(storer.Users, oracle.NumOfUsers)
 }
 
-// TestOnlineOrganizationDownload Tests the download of known and fixed GitHub organization
-func TestOnlineOrganizationDownload(t *testing.T) {
+// TestOfflineOrganizationDownload Tests a large organization by replaying recorded responses
+func (suite *DownloaderTestSuite) TestOfflineOrganizationDownload() {
+	t := suite.T()
+	reqResp := make(map[string]string)
+	// Load the recording
+	suite.NoError(loadReqResp("../testdata/organization_src-d_2019-10-08.gob", reqResp), "Failed to read the offline recordings")
+	// Setup the downloader with RoundTrip functionality.
+	// Not using the NewStdoutDownloader initialization because it overides the transport
+	storer := &testutils.Memory{}
+	downloader := &Downloader{
+		storer: storer,
+		client: githubv4.NewClient(&http.Client{
+			Transport: RoundTripFunc(func(req *http.Request) *http.Response {
+				// consume request body
+				savecl := req.ContentLength
+				bodyBytes, _ := ioutil.ReadAll(req.Body)
+				defer req.Body.Close()
+				// recreate request body
+				req.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+				req.ContentLength = savecl
+				data := reqResp[string(bodyBytes)]
+				return &http.Response{
+					StatusCode: 200,
+					Body:       ioutil.NopCloser(bytes.NewBufferString(data)),
+					Header:     make(http.Header),
+				}
+			})}),
```

Comment:
this part is equal in several tests, could we move it to some external function like `getDownloader` or so?